algorithms/nautilus/grammartec/context.cpp

```diff
@@ -0,0 +1,495 @@
+ * @brief Get minimum length for nonterminal
+ * @param (nt) Nonterminal ID
+ * throw std::out_of_range Nonterminal ID is invalid
+ * @return Minimum length for @p nt
+ */
+size_t Context::GetMinLenForNT(const NTermID& nt) const {
+  return _nts_to_min_size.at(nt);
+}
+
+/**
+ * @fn
+ * Register a new nonterminal symbol and get its ID
+ * @brief Register a new nonterminal symbol
+ * @param (nt) String of a nonterminal symbol
+ * @return Nonterminal ID of @p nt
+ *
+ * @details This method looks up a nonterminal symbol.
+ *          If there exists one, the corresponding nonterminal ID
+ *          is returned.
+ *          Otherwise this method registers the new nonterminal symbol
+ *          and returns a new ID.
+ */
+NTermID Context::AquireNTID(const std::string& nt) {
+  NTermID next_id(_nt_ids_to_name.size()); // New NTermID
+
+  NTermID& id = _names_to_nt_id.find(nt) == _names_to_nt_id.end()
+    ? (_names_to_nt_id[nt] = next_id) // not exists
+    : _names_to_nt_id[nt];            // exists
+
+  if (_nt_ids_to_name.find(id) == _nt_ids_to_name.end())
+    _nt_ids_to_name[id] = nt;
+
+  return id;
+}
+
+/**
+ * @fn
+ * Return nonterminal ID referenced by a nonterminal symbol
+ * @brief Lookup NTermID by nonterminal symbol
+ * @param (nt) String of a nonterminal symbol
+ * @throw std::out_of_range Nonterminal symbol does not exist
+ * @return Nonterminal referenced by @p nt
+ */
+const NTermID& Context::NTID(const std::string& nt) const {
+  return _names_to_nt_id.at(nt);
+}
+
+/**
+ * @fn
+ * Add a new plain rule to this context.
+ * @brief Add a new rule
+ * @param (nt) Nonterminal symbol
+ * @param (format) Format string (expression)
+ * @return Registered rule ID
+ */
+RuleID Context::AddRule(const std::string& nt, const std::string& format) {
+  RuleID rid(_rules.size()); // New rule ID
+  const NTermID& ntid = AquireNTID(nt);
+
+  /* Register this rule */
+  _rules.emplace_back(*this, nt, format);
+
+  if (_nts_to_rules.find(ntid) == _nts_to_rules.end())
+    _nts_to_rules[ntid] = {};
+  _nts_to_rules[ntid].emplace_back(rid);
+
+  return rid;
+}
+
+/**
+ * @fn
+ * Calculate the number of options for a rule.
+ * @brief Calculate number of options
+ * @param (r) RuleID
+ * @return Number of options for @p r
+ */
+size_t Context::CalcNumOptionsForRule(const RuleID& r) const {
+  size_t res = 1;
+
+  for (const NTermID& nt_id: GetRule(r).Nonterms()) {
+    size_t v = _nts_to_num_options.find(nt_id) == _nts_to_num_options.end()
+      ? 1
+      : _nts_to_num_options.at(nt_id);
+
+    if (__builtin_mul_overflow(res, v, &res)) {
+      /* Saturate instead of overflow */
+      res = std::numeric_limits<size_t>::max();
+    }
+  }
+
+  return res;
+}
+
+/**
+ * @fn
+ * @brief Calculate number of options
+ */
+void Context::CalcNumOptions() {
+  for (auto& elem: _nts_to_rules) {
+    if (_nts_to_num_options.find(elem.first) == _nts_to_num_options.end()) {
+      _nts_to_num_options[elem.first] = _rules.size();
```

Comment:
Probably this should be `elem.second.size()`.

Reply:
Fixed in [80717f4](https://github.com/fuzzuf/fuzzuf/pull/23/commits/80717f42d2b634a8c9d5b242aca7a897c2a78305)